templates/language.rs

```diff
@@ -0,0 +1,129 @@
+        match language {
+            {% for language, _ in languages -%}
+                Language::{{ language }} => tokei::LanguageType::{{ language }},
+            {% endfor %}
+        }
+    }
+}
+
+impl Language {
+    pub fn get_ascii_art(&self) -> &'static str {
+        match self {
+            {% for language, attrs in languages -%}
+                Language::{{ language }} => "{{ attrs.ascii | addslashes }}",
+            {% endfor %}
+        }
+    }
+
+    pub fn get_colors(&self, true_color: bool) -> Vec<DynColors> {
+        let colors = match self {
+            {% for language, attrs in languages -%}
+                Language::{{ language }} => Colors {
+                    basic_colors: vec![{%- for color in attrs.colors.ansi -%}Ansi(AnsiColors::{{ color | title }}),{% endfor %}],
+                    true_colors: {% if attrs.colors.rgb %}Some(vec![{%- for rgb in attrs.colors.rgb -%}Rgb({{ rgb | nth(n=0) }}, {{ rgb | nth(n=1) }}, {{ rgb | nth(n=2) }}),{% endfor %}]){% else %}None{% endif %},
+                },
+            {% endfor %}
+        };
+        match colors.true_colors {
+            Some(true_colors) if true_color => true_colors,
+            _ => colors.basic_colors,
+        }
+    }
+
+    pub fn get_type(&self) -> LanguageType {
+        match self {
+            {% for language, attrs in languages -%}
+                Language::{{ language }} => LanguageType::{{ attrs.type | title }},
+            {% endfor %}
+        }
+    }
+
+    pub fn get_circle_color(&self) -> DynColors {
+        match self {
+            {% for language, attrs in languages -%}
+                Language::{{ language }} => Rgb({{ attrs.colors.chip | nth(n=0) }}, {{ attrs.colors.chip | nth(n=1) }}, {{ attrs.colors.chip | nth(n=2) }}),
+            {% endfor %}
+        }
+    }
+}
+
+{% for language, attrs in languages -%}
+    {% if attrs.colors.rgb %}
+        {% set ansi_length = attrs.colors.ansi | length %}
+        {% set rgb_length = attrs.colors.rgb | length %}
+        {% if ansi_length != rgb_length %}
+            compile_error!("{{ language }}: ansi and rgb colors must be the same length");
+        {% endif %}
+    {% endif %}
+{% endfor %}
+
+{% set max_width = 40 %}
+{# NOTE Permitting trailing newline #}
+{% set max_height = 26 %}
+
+
+{% for language, attrs in languages -%}
+    {% set lines = attrs.ascii | split(pat="\n") %}
+    {% set height = lines | length %}
+    {% if height > max_height %}
+        compile_error!("{{ language }}: ascii art must have {{ max_height - 1 }} or less lines, has {{ height }}");
+    {% endif %}
+
+    {% for line in lines %}
+        {% set cleaned_line = line | strip_color_indices %}
+        {% set width = cleaned_line | length %}
+        {% if width > max_width %}
+            compile_error!("{{ language }}: ascii art line {{ loop.index }} must be {{ max_width }} or less characters wide");
+        {% endif %}
+    {% endfor %}
+{% endfor %}
```

Comment:
I'm not sure why but if you look at the generated `language.rs` by tera in the /target/debut/onefetch-*/out folder, this part of the template results in thousands of empty lines.

Besisdes, IMO these should be unit tests not compilation errors. We should maybe find a way to put them back in the original `language.rs` file.

Reply:
Yeah, the empty lines are from the `if` statements never being entered. E.g.
```
{# 3 empty lines #}
{% for n in [1,2,3] %}
  {% if n > 4 %}println!("Hello!"){% endif %}
{% endfor %}
```
IMO we don't need to worry too much about how pretty the generated code is: generated code isn't really intended to be human-readable.

For compile errors vs tests, the benefit of compile errors is that you can't even run onefetch if the values are invalid. Additionally, IMO it's preferable to assert the validity of values at compile-time instead of a runtime (in this case the test runtime). IMO tests should really be for behavior, not validating internal values. As a reductionist example, it's kind of like doing this:
```rust
/// A number between 0 and 255
const SMALL_NUMBER: usize = make_number!();

#[test]
fn number_is_small() {
    assert!(SMALL_NUMBER < 256);
}
```

instead of this:
```rust
/// A number between 0 and 255
const SMALL_NUMBER: u8 = make_number!();
```
Making them compile errors also should result in a faster CI, since the presence of a `compile_error!` macro in the generated code will make compilation fail pretty quickly.

Reply:
Agreed 👍 

> Yeah, the empty lines are from the if statements never being entered. E.g.

Maybe add some `-%`  to reduce the clutter.